### PR TITLE
Fixed Integ Test To Work With Multiple Plugin and Cluster Status

### PR DESCRIPTION
**Pull request by owaiskazi19**
### Description
Fixed integ test to work for multiple plugin and removed the green status check for local test cluster
 
### Issues Resolved
https://github.com/opensearch-project/opensearch-build/issues/473 and https://github.com/opensearch-project/opensearch-build/issues/500
 
### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: bundle-workflow/src/test_workflow/config/test_manifest.yml [hunk 1]
```diff
@@ -12,4 +12,56 @@ components:
...
+        - job-scheduler
+        - alerting
+      test-configs:
+        - with-security
+        - without-security
+    bwc-test:
+      dependencies:
+      test-configs:
+       - with-security
+        - without-security
```
Comment (saratvemulapalli) [inline bundle-workflow/src/test_workflow/config/test_manifest.yml:66]: nit: indentation
```suggestion
        - with-security
        - without-security
```

File: bundle-workflow/src/test_workflow/integ_test/local_test_cluster.py [hunk 2]
```diff
@@ -92,7 +92,7 @@ def wait_for_service(self):
                 logging.info(f"Pinging {url} attempt {attempt}")
                 response = requests.get(url, verify=False, auth=("admin", "admin"))
                 logging.info(f"{response.status_code}: {response.text}")
-                if response.status_code == 200 and '"status":"green"' in response.text:
```
Comment (saratvemulapalli) [inline bundle-workflow/src/test_workflow/integ_test/local_test_cluster.py:95]: Check if the cluster is yellow/green. 
We do not want to proceed if the cluster is red.

File: bundle-workflow/src/test_workflow/integ_test/local_test_cluster.py [hunk 3]
```diff
@@ -92,7 +92,7 @@ def wait_for_service(self):
...
+                if response.status_code == 200:
                     logging.info("Cluster is green")
```
Comment (saratvemulapalli) [inline bundle-workflow/src/test_workflow/integ_test/local_test_cluster.py:96]: Please fix this comment and other places within this code where "green" is used.

Comment (owaiskazi19) [reply bundle-workflow/src/test_workflow/integ_test/local_test_cluster.py:96]: That's right should fix this. Any suggestion you got for the comment?

File: bundle-workflow/src/test_workflow/integ_test/local_test_cluster.py (hunk 2 above)
Comment (owaiskazi19) [reply bundle-workflow/src/test_workflow/integ_test/local_test_cluster.py:95]: I suppose the `status_code` will be 200 only when the cluster is green/yellow. Correct me if I am wrong here

File: bundle-workflow/src/test_workflow/config/test_manifest.yml [hunk 4]
```diff
@@ -12,4 +12,56 @@ components:
       test-configs:
         - with-security
         - without-security
+  - name: anomaly-detection
+    integ-test:
+      dependencies:
+        - job-scheduler
+        - alerting
```
Comment (VachaShah) [inline bundle-workflow/src/test_workflow/config/test_manifest.yml:19]: `job-scheduler` and `alerting` won't be dependencies for all plugins. Since we are not using this object yet, may be you could keep empty for all other plugins or add the correct dependencies for each plugin to avoid confusion.

File: bundle-workflow/src/test_workflow/integ_test/local_test_cluster.py (hunk 2 above)
Comment (setiah) [reply bundle-workflow/src/test_workflow/integ_test/local_test_cluster.py:95]: it can be 200 and red

File: bundle-workflow/src/test_workflow/integ_test/local_test_cluster.py (hunk 3 above)
Comment (setiah) [reply bundle-workflow/src/test_workflow/integ_test/local_test_cluster.py:96]: Cluster is green --> Service is available

File: bundle-workflow/src/test_workflow/integ_test/local_test_cluster.py (hunk 2 above)
Comment (saratvemulapalli) [reply bundle-workflow/src/test_workflow/integ_test/local_test_cluster.py:95]: Correct status code will be 200, just verify `response.text` has green/yellow.

File: bundle-workflow/src/test_workflow/integ_test/local_test_cluster.py (hunk 3 above)
Comment (saratvemulapalli) [reply bundle-workflow/src/test_workflow/integ_test/local_test_cluster.py:96]: May be "Cluster is healthy"/"Cluster is ready for the test".

File: bundle-workflow/src/test_workflow/config/test_manifest.yml [hunk 5]
```diff
@@ -12,4 +12,56 @@ components:
...
+    bwc-test:
+      dependencies:
+      test-configs:
+        - with-security
+        - without-security
+  - name: sql
+    integ-test:
+      dependencies:
+        - job-scheduler
+        - alerting
```
Comment (setiah) [inline bundle-workflow/src/test_workflow/config/test_manifest.yml:45]: (outside of scope of this PR) we should remove these `dependencies` in a separate PR and remove them from test_manifest schema. This will not be used.

File: bundle-workflow/src/test_workflow/config/test_manifest.yml (hunk 4 above)
Comment (setiah) [reply bundle-workflow/src/test_workflow/config/test_manifest.yml:19]: I think we should just remove them to keep things simple https://github.com/opensearch-project/opensearch-build/pull/512#discussion_r710581526. The dependency management logic needs to be handled via maven eventually.

File: bundle-workflow/src/test_workflow/integ_test/local_test_cluster.py (hunk 2 above)
Comment (owaiskazi19) [reply bundle-workflow/src/test_workflow/integ_test/local_test_cluster.py:95]: Changed the message to `Service is available`

File: bundle-workflow/src/test_workflow/config/test_manifest.yml (hunk 5 above)
Comment (owaiskazi19) [reply bundle-workflow/src/test_workflow/config/test_manifest.yml]: Removed the `dependencies` and removed from schema as well

File: bundle-workflow/src/test_workflow/integ_test/local_test_cluster.py (hunk 3 above)
Comment (owaiskazi19) [reply bundle-workflow/src/test_workflow/integ_test/local_test_cluster.py]: Changed the message to `Service is available`

File: bundle-workflow/src/test_workflow/config/test_manifest.yml [hunk 6]
```diff
@@ -1,13 +1,51 @@
...
+      test-configs:
+        - with-security
+        - without-security
+  - name: asynchronous-search
+    integ-test:
+      test-configs:
+        - with-security
+        - without-security
+    bwc-test:
+      dependencies:
```
Comment (VachaShah) [inline bundle-workflow/src/test_workflow/config/test_manifest.yml:28]: *Nit*: This section can be removed as well. You can do it as part of another PR though. Don't want to block the approval for it.

Comment (owaiskazi19) [reply bundle-workflow/src/test_workflow/config/test_manifest.yml:32]: If bwc-test doesn't uses `dependencies`. I can remove it

**Review by VachaShah — DISMISSED** [review]

File: bundle-workflow/src/run_integ_test.py [hunk 7]
```diff
@@ -41,13 +41,14 @@ def pull_common_dependencies(work_dir, build_manifest):
...
-                os.path.join(work_dir, component.name),
+                os.path.join(work_dir, 'dependencies', component.name),
             )
 
 
 # TODO: replace with DependencyProvider - https://github.com/opensearch-project/opensearch-build/issues/283
 def sync_dependencies_to_maven_local(work_dir, manifest_build_ver):
-    os.chdir(work_dir + "/OpenSearch")
+    dependencies_dir = os.path.join(work_dir, 'dependencies')
+    os.chdir(dependencies_dir + "/OpenSearch")
```
Comment (dblock) [inline bundle-workflow/src/run_integ_test.py:51]: Change `run` to `check_output`, which incidentally supports `cwd=` for the directory in which it runs.

**Review by dblock — COMMENTED** [review]
Why are we removing `dependencies`? I thought that was a pretty good idea to keep things more data-driven/configurable. An alternative could be a custom script found via ScriptFinder that gets executed for a component that needs something. Either way hard-coding copying plugin files to hack our way to a working system doesn't seem like a great idea - we'll have to fix it again for OpenSearch Dashboards.

File: bundle-workflow/src/run_integ_test.py (hunk 7 above)
Comment (setiah) [reply bundle-workflow/src/run_integ_test.py:51]: +1

**Comment by setiah** [comment]
> Why are we removing `dependencies`? I thought that was a pretty good idea to keep things more data-driven/configurable. An alternative could be a custom script found via ScriptFinder that gets executed for a component that needs something. Either way hard-coding copying plugin files to hack our way to a working system doesn't seem like a great idea - we'll have to fix it again for OpenSearch Dashboards.

Hey @dblock, the dependencies defined in the config file are not being used today. They were planned to be used by DependencyInstaller, which is wip. So it will be added back later, in the right format, w/ dependency installer once that is integrated.

**Comment by setiah** [comment]
> > Why are we removing `dependencies`? I thought that was a pretty good idea to keep things more data-driven/configurable. An alternative could be a custom script found via ScriptFinder that gets executed for a component that needs something. Either way hard-coding copying plugin files to hack our way to a working system doesn't seem like a great idea - we'll have to fix it again for OpenSearch Dashboards.
> 
> Hey @dblock, the dependencies defined in the config file are not being used today. They were planned to be used by DependencyInstaller, which is wip. So it will be added back later, in the right format, w/ dependency installer once that is integrated.

A better way IMO would be if we can publish these maven artifacts to an internal s3 repo, the dependency management logic for plugins can be handled by gradle itself. This would be a much better solution given the plugin dependencies will keep evolving and the build repo will not be able to keep up. Having a consistent dependency management solution via maven should be the right way in the long run

**Comment by dblock** [comment]
@setiah I would just keep the manifest that lists dependencies instead of hard-coding the list in code. The code itself can loop through dependencies and run whatever command it needs to publish these/build/whatever. This is what's done in manifest checks.

**Comment by setiah** [comment]
> @setiah I would just keep the manifest that lists dependencies instead of hard-coding the list in code. The code itself can loop through dependencies and run whatever command it needs to publish these/build/whatever. This is what's done in manifest checks.

Had a quick chat with @dblock on this. The PR looks good to merge. In the subsequent revisions, we would change the dependency management logic to copy all maven dependencies from s3 to maven local, instead of copying plugin specific dependencies. This would simplify the logic further.

In future, a good optimization would be to make the maven path configurable in the integtest.sh. This would allow using a custom maven repo while running integ tests and automatically manage dependencies for a plugin with the help of gradle.

**Comment by owaiskazi19** [comment]
@dblock @setiah I have added back the `dependencies` as it's not the part of this PR. If it's required we can create a separate issue and address it in the future PR.

**Comment by owaiskazi19** [comment]
> > @setiah I would just keep the manifest that lists dependencies instead of hard-coding the list in code. The code itself can loop through dependencies and run whatever command it needs to publish these/build/whatever. This is what's done in manifest checks.
> 
> Had a quick chat with @dblock on this. The PR looks good to merge. In the subsequent revisions, we would change the dependency management logic to copy all maven dependencies from s3 to maven local, instead of copying plugin specific dependencies. This would simplify the logic further.
> 
> In future, a good optimization would be to make the maven path configurable in the integtest.sh. This would allow using a custom maven repo while running integ tests and automatically manage dependencies for a plugin with the help of gradle.

I missed this. I'll address it in the next PR

File: bundle-workflow/src/test_workflow/config/test_manifest.yml [hunk 8]
```diff
@@ -8,6 +8,69 @@ components:
...
+        - job-scheduler
+        - alerting
+      test-configs:
+        - with-security
+        - without-security
+  - name: anomaly-detection
+    integ-test:
+      dependencies:
+        - job-scheduler
+        - alerting
```
Comment (saratvemulapalli) [inline bundle-workflow/src/test_workflow/config/test_manifest.yml:21]: Why are all the plugins dependent on alerting and job-scheduler?

Comment (owaiskazi19) [reply bundle-workflow/src/test_workflow/config/test_manifest.yml:21]: As per the discussion https://github.com/opensearch-project/opensearch-build/pull/512#issuecomment-921975871.  I'll address it in the next PR.

Comment (saratvemulapalli) [reply bundle-workflow/src/test_workflow/config/test_manifest.yml:21]: Sure. I am fine addressing this in a quick follow up PR. If so please create an issue so we have a track.

But always, as best practices: 
- Don't commit code we are not planning to use.
- Commit incorrect configuration which is un-usable.

**Review by saratvemulapalli — APPROVED** [review]
Dropped a comment, other than that rest of the changes look good to me.
Thanks @owaiskazi19 !

File: bundle-workflow/src/test_workflow/config/test_manifest.yml (hunk 8 above)
Comment (owaiskazi19) [reply bundle-workflow/src/test_workflow/config/test_manifest.yml:21]: Yeah, that makes sense. Thanks for the new learning. I have create an issue to track this down: https://github.com/opensearch-project/opensearch-build/issues/521

**Review by VachaShah — APPROVED** [review]
LGTM!
